### ISSUE TITLE
CI: Conformance IPSec E2E, logs spammed with HUBBLE_RING_BUFFER lost events

### ISSUE THREAD

**Issue by squeed** [cilium/cilium#34981]
## CI failure

In an [innocuous PR](https://github.com/cilium/cilium/pull/34870#) on main, one of the Conformance IPSec E2E [runs failed](https://github.com/cilium/cilium/actions/runs/10882926526/job/30194988581#logs). The logs were filled with the line

```
⚠️ hubble-flows-cilium-8nlvr: failed to collect hubble flows for "cilium-8nlvr" in namespace "kube-system": error with exec request (pod=kube-system/cilium-8nlvr, container=cilium-agent): command terminated with exit code 1: time=2024-09-16T12:11:31.040Z level=DEBUG msg="connected to Hubble API" server=server
time=2024-09-16T12:11:31.040Z level=DEBUG msg="Sending GetFlows request" request=number:1000000
unknown response type: &{LostEvents:source:HUBBLE_RING_BUFFER num_events_lost:1}
unknown response type: &{LostEvents:source:HUBBLE_RING_BUFFER num_events_lost:1}
unknown response type: &{LostEvents:source:HUBBLE_RING_BUFFER num_events_lost:1}
unknown response type: &{LostEvents:source:HUBBLE_RING_BUFFER num_events_lost:1}
(repeats thousands of times)
```

I saw this once more today, but I can't find the link right now.

**Comment by bimmlerd** [comment]
@glrf and I looked into this one rainy morning as I hit this in some PR too - seems like this is occurring more often than we thought, so maybe it's worth investigating a bit more.

**Comment by Trump-last** [comment]
On Cilium version 1.15.6, we discovered this issue too. After reviewing the source code, I believe that Cilium should not push all lost events into the event queue without aggregation. Under high-pressure conditions, this could starve normal events from being written to the event queue (I don’t understand why we aggregated lost events for logging output but didn’t apply the same approach when writing to the event queue).

Additionally, the filtering does not occur when writing to the event queue but rather when reading from the Ring buffer. I find this approach unreasonable because it allows a large number of useless traffic events to occupy significant space and rapidly fill up the event queue. Shouldn’t we control which events are written to the event queue right from the source?

Lost events are flooding my logs, to the point where I can't even retrieve normal flow data through the GetFlows() gRPC call. H E L P!
![Image](https://github.com/user-attachments/assets/956a4bfc-e69b-4da3-bf08-8be547ee183a)
![Image](https://github.com/user-attachments/assets/00827a9e-ff24-4c7b-9232-3b948038d102)
```
func (c *consumer) sendNumLostEvents() {
	c.lostLock.Lock()
	defer c.lostLock.Unlock()
	// check again, in case multiple
	// routines contended the lock
	if c.numEventsLost == 0 {
		return
	}

	numEventsLostNotification := &observerTypes.MonitorEvent{
		UUID:      uuid.New(),
		Timestamp: time.Now(),
		NodeName:  nodeTypes.GetAbsoluteNodeName(),
		Payload: &observerTypes.LostEvent{
			Source:        observerTypes.LostEventSourceEventsQueue,
			NumLostEvents: c.numEventsLost,
		},
	}
	select {
	case c.observer.GetEventsChannel() <- numEventsLostNotification:
		// We now now safely reset the counter, as at this point have
		// successfully notified the observer about the amount of events
		// that were lost since the previous LostEvent message
		c.numEventsLost = 0
	default:
		// We do not need to bump the numEventsLost counter here, as we will
		// try to send a new LostEvent notification again during the next
		// invocation of sendEvent
	}
}

// sendEvent enqueues an event in the observer. If this is not possible, it
// keeps a counter of lost events, which it will regularly try to send to the
// observer as well
func (c *consumer) sendEvent(event *observerTypes.MonitorEvent) {
	if c.numEventsLost > 0 {
		c.sendNumLostEvents()
	}
	select {
	case c.observer.GetEventsChannel() <- event:
	default:
		c.countDroppedEvent()
	}
}

// countDroppedEvent logs that the events channel is full
// and counts how many messages it has lost.
func (c *consumer) countDroppedEvent() {
	c.lostLock.Lock()
	defer c.lostLock.Unlock()
	if c.numEventsLost == 0 && c.logLimiter.Allow() {
		c.observer.GetLogger().WithField("related-metric", "hubble_lost_events_total").
			Warning("hubble events queue is full: dropping messages; consider increasing the queue size (hubble-event-queue-size) or provisioning more CPU")
	}
	c.numEventsLost++
	metrics.LostEvents.WithLabelValues(strings.ToLower(flowpb.LostEventSource_OBSERVER_EVENTS_QUEUE.String())).Inc()
}
```

**Comment by Trump-last** [comment]
The two issues mentioned above can lead to a disastrous scenario in high-traffic clusters:

Here’s how it happens:

A flow event that should have been filtered during read time overflows before entering the event queue.
It then gets converted into a lost event—but filtering rules don’t apply to lost events.
As a result, during retrieval, this event escapes filtering entirely and is returned to the caller via GetFlows().

The end result?

What was supposed to be filtered out now leaks back as a lost event—and the poor caller gets flooded with them.
When this occurs at scale, it’s an absolute disaster...

**Comment by Trump-last** [comment]
![Image](https://github.com/user-attachments/assets/179f69c9-5d03-44b4-acaf-117fbcc84981)
![Image](https://github.com/user-attachments/assets/18f3ddfc-eba4-466c-b0af-d4da799bca99)
![Image](https://github.com/user-attachments/assets/364a2f11-b703-40ec-b08c-a6470c5c58b7)
This is the log information we obtained through Prometheus metrics. The `flow` field represents the number of normally received flow messages, while the `lost` field indicates the number of lost events.

We can see that from 03:22:29 to 03:22:59, the normal flow data reception rate was 1,500 messages per second, which was a traffic peak during our testing.

Subsequently, from 03:23:29 to 03:23:59, an overflow event occurred, likely due to the impact of the preceding traffic peak, causing the number of lost events to surge rapidly.

Then, we observe that between 03:23:29 and 03:25:00, the rate at which overflow events were generated far exceeded expectations, reaching 30,000 per second, while the normal flow event generation rate dropped to just a few hundred per second—an abnormal situation.

This indicates two issues:
1. Many flow events that should have been filtered were instead converted into lost events and returned.
2. Many flow events that should have been returned were lost due to overflow, as the buffer was flooded with lost events.

Since lost events need to be returned to the caller, they introduce additional processing overhead (especially if the original flow events should have been filtered and ignored). This further slows down the consumer's event processing speed, increasing the likelihood of overflow. Under the compounding effect of these two factors, an overflow storm begins...

![Image](https://github.com/user-attachments/assets/6ef25bbb-7fd5-4516-abcd-3d9c6b3c1507)

**Comment by kaworu** [comment]
@Trump-last thank you for the investigation. IIRC the intent was for lost events to be sent once the traffic surge stopped and the Hubble observer (i.e. the monitor consumer) "recovered" from it. However, as you pointed out, in practice it doesn't work that way.

I think the lost event accounting has to change, because we get too many `&{LostEvents:source:OBSERVER_EVENTS_QUEUE num_events_lost:1}` where we would like less messages with higher `num_events_lost` numbers (i.e. more aggressive aggregation of lost events). So here is a high level idea:
1. the first time we fail to send an event, set the **lost events counter** to `1` and  set an **lost events timestamp**.
2. every time we fail to send an event again in a defined time window (e.g. 3s), increase the **lost events counter**.
3. once we reached the end of the lost events window (i.e. 3s after **lost events timestamp** of the first lost event), send a lost event message to the observer with the **lost events counter** and time window (i.e. "`n` events were lost between `2025-06-10 13:57:42 +0200` and `2025-06-10 13:57:45 +0200`"

That way we guarantee to send at most 1 lost event each time window (e.g. 3s but could even be configurable on the server side).

**Comment by rolinh** [comment]
@kaworu I agree with what you propose and iirc, this is how it was meant to be implemented.

**Comment by gandro** [comment]
@kaworu I think your proposal works. One minor downside is that if the client disconnects before end of the window then they never know how many events they lost. But it likely doesn't matter in practice.

**Comment by kaworu** [comment]
> One minor downside is that if the client disconnects before end of the window then they never know how many events they lost

Correct, although I guess to some extend this case can happen with the current implementation when the client disconnects while monitor is failing to send the lost event count in `sendNumLostEvents`?

In any case, we can start with a relatively short window of 1s: the reasoning being that it should be still long enough to "let events go through" (i.e. avoid the case where the consumer is flooded with lost events described above) but short enough to be practical in interactive usage (in the typical `hubble observe` CLI invocation against Hubble Relay there is a [1s buffer for sorting](https://github.com/cilium/cilium/blob/2732b5253beef350dd89216813d4e4da1180e8c7/pkg/hubble/relay/defaults/defaults.go#L35-L37) already). What do you think?

**Comment by Trump-last** [comment]
> [@Trump-last](https://github.com/Trump-last) thank you for the investigation. IIRC the intent was for lost events to be sent once the traffic surge stopped and the Hubble observer (i.e. the monitor consumer) "recovered" from it. However, as you pointed out, in practice it doesn't work that way.
> 
> I think the lost event accounting has to change, because we get too many `&{LostEvents:source:HUBBLE_RING_BUFFER num_events_lost:1}` where we would like less messages with higher `num_events_lost` numbers (i.e. more aggressive aggregation of lost events). So here is a high level idea:
> 
> 1. the first time we fail to send an event, set the **lost events counter** to `1` and  set an **lost events timestamp**.
> 2. every time we fail to send an event again in a defined time window (e.g. 3s), increase the **lost events counter**.
> 3. once we reached the end of the lost events window (i.e. 3s after **lost events timestamp** of the first lost event), send a lost event message to the observer with the **lost events counter** and time window (i.e. "`n` events were lost between `2025-06-10 13:57:42 +0200` and `2025-06-10 13:57:45 +0200`"
> 
> That way we guarantee to send at most 1 lost event each time window (e.g. 3s but could even be configurable on the server side).

@kaworu
Yes, I believe reducing the volume of lost event messages is crucial because, based on recent traffic log observations, we've found that `&{LostEvents:source:HUBBLE_RING_BUFFER num_events_lost:1}` can significantly impact the performance of Hubble Flow gRPC callers.

In our usage scenario, a vicious cycle can occur: once our gRPC clients receive a large number of `&{LostEvents:source:HUBBLE_RING_BUFFER num_events_lost:1}` events, it exacerbates the problem. This is because, in our setup, we log these overflow messages. If I understand correctly, the root cause of `&{LostEvents:source:HUBBLE_RING_BUFFER num_events_lost:1}` is that the consumer side cannot keep up with the producer side's speed. If the consumer then receives a large volume of such messages, its performance degrades further, leading to even more overflow from the producer side (Cilium agent).

From our log observations, in large-scale, high-traffic clusters, these overflow events often outnumber normal flow events by dozens of times (in our case, we only focus on north-south traffic). This means that processing such overflow events often incurs a significant performance penalty.

**Comment by Trump-last** [comment]
> > One minor downside is that if the client disconnects before end of the window then they never know how many events they lost
> 
> Correct, although I guess to some extend this case can happen with the current implementation when the client disconnects while monitor is failing to send the lost event count in `sendNumLostEvents`?
> 
> In any case, we can start with a relatively short window of 1s: the reasoning being that it should be still long enough to "let events go through" (i.e. avoid the case where the consumer is flooded with lost events described above) but short enough to be practical in interactive usage (in the typical `hubble observe` CLI invocation against Hubble Relay there is a [1s buffer for sorting](https://github.com/cilium/cilium/blob/2732b5253beef350dd89216813d4e4da1180e8c7/pkg/hubble/relay/defaults/defaults.go#L35-L37) already). What do you think?

@kaworu 
In our most recent log observation, we discovered that if a node experiences performance bottlenecks—such as certain tasks maxing out the CPU on that node or brief network fluctuations—the Cilium agent, which serves as the server sending traffic records to Hubble Relay, begins to emit a large number of `&{LostEvents:source:HUBBLE_RING_BUFFER num_events_lost:1}` events. The most severe issue is that this situation triggers the "vicious cycle" I mentioned earlier: the agent starts sending `&{LostEvents:source:HUBBLE_RING_BUFFER num_events_lost:1}` events → the gRPC client becomes overwhelmed → even more `&{LostEvents:source:HUBBLE_RING_BUFFER num_events_lost:1}` events are generated.

In large-scale, high-traffic clusters, the Cilium agent currently lacks mitigation or recovery mechanisms for this problem. Even a few seconds of network fluctuations can lead to an unrecoverable state. I believe this is a critical issue that warrants attention and requires resolution.

**Comment by kaworu** [comment]
@Trump-last thank you for the feedback. From your reports I see two related but distinct issues:

1. `LostEvents:source:OBSERVER_EVENTS_QUEUE` are generated globally at the cilium monitor consumer level. Theses are happening more than wished, but they are (sometimes) aggregated, i.e. there are instances where `num_events_lost > 1`.
2. `LostEvents:source:HUBBLE_RING_BUFFER` are generated per gRPC client attempting to read the Hubble ring buffer. Theses are never aggregated, as [the current codepath](https://github.com/cilium/cilium/blob/bcce70c6314653cbea99148c6f77995fc216b7e8/pkg/hubble/container/ring.go#L222-L232) always return them with `num_events_lost = 1`.

> From our log observations, in large-scale, high-traffic clusters, these overflow events often outnumber normal flow events by dozens of times (in our case, we only focus on north-south traffic)

Out of curiosity, do you have a way to further identify theses lost events and count the number of `OBSERVER_EVENTS_QUEUE` vs `HUBBLE_RING_BUFFER`?

> In large-scale, high-traffic clusters, the Cilium agent currently lacks mitigation or recovery mechanisms for this problem. Even a few seconds of network fluctuations can lead to an unrecoverable state. I believe this is a critical issue that warrants attention and requires resolution.

Yes, I agree that in high pressure scenarios Hubble should not imper the Cilium agent stability. I've discussed with @devodev about the specific and he's currently working on a solution.

While [PR#41163](https://github.com/cilium/cilium/pull/41163) should help with your issue, it's worth pointing out that Hubble Relay is a single gRPC API consumer and its design isn't really suited for collecting all the flows from all the nodes in a large-scale, high traffic cluster. Our own Hubble Relay testing so far has been limited to 128 nodes and ~40k flows/sec (dating back several years to Relay v1.8 though, so should be taken with a grain of salt).
